numpy/lib/function_base.py

```diff
@@ -1320,6 +1320,258 @@ def gradient(f, *varargs, **kwargs):
+    else:
+        if False in [np.shape(v[0]) == np.shape(v[i]) for i in range(len(v))]:
+            raise ValueError("Not all components of input are the same size")
+        else:
+	    N = N[0] 
+	# If all vector field components are the same shape,
+	#N becomes the number of dimensions
+    
+    axes = tuple(range(N))
+    outvals=np.zeros(np.shape(v_c)) #Initialize output
+    
+    for i,ax in enumerate(axes):
+        outvals += np.gradient(v[ax], *varargs, axis=ax,edge_order=0)
+	
+    return outvals
+		
+		
+def curl(v,*varargs,**kwargs):
+    """
+    Return the curl of a 3-D vector field.
+
+    The curl is computed using second order accurate central differences
+    in the interior and either first differences or second order accurate
+    one-sides (forward or backwards) differences at the boundaries. The
+    returned gradient hence has the same shape as the input array.
+    
+    Parameters
+    ----------
+    v : list of numpy arrays
+        Each of these array is the N-dimensional projection 
+        of the vector field on the corresponding axis. 
+    
+    varargs : scalar or list of scalar, optional
+        N scalars specifying the sample distances for each dimension,
+        i.e. `dx`, `dy`, `dz`, ... Default distance: 1.
+        single scalar specifies sample distance for all dimensions.
+        if `axis` is given, the number of varargs must equal the number of axes.
+    edge_order : {1, 2}, optional
+        Gradient is calculated using N\ :sup:`th` order accurate differences
+        at the boundaries. Default: 1.
+        
+    axis : None or int or tuple of ints, optional
+        Gradient is calculated only along the given axis or axes
+        The default (axis = None) is to calculate the gradient for 
+        all the axes of the input array.
+        axis may be negative, in which case it counts from the last to the first axis.
+        
+    Returns
+    -------
+    curl : list of numpy arrays
+        The output corresponds to the curl of the input 3-D vector field.
+        This means that the output has the form 
+        (dAz/dy-dAy/dz, dAx/dz-dAz/dx, dAy/dx-dAx/dy) of an input vector 
+        field A = (Ax, Ay, Ax).
+    
+    Example
+    -------
+    The following example shows in matplotlib how the vector field (0,0,x*y) has the 
+    correct value, nonzero only in the third component of the resulting vector field
+    after applying curl
+    
+    >>> import matplotlib.pylab as plt
+    >>> import numpy as np
+    >>> X,Y,Z=np.mgrid[0:200,0:200,0:200]
+    >>> a0=X*Y
+    >>> a1=np.zeros(np.shape(X))
+    >>> a2=np.zeros(np.shape(X))
+    >>> a=[a0,a1,a2]
+    >>> [cx,cy,cz]=np.curl(a)
+    >>> plt.imshow(cx[:,:,1])
+    >>> plt.colorbar()
+    >>> plt.show()
+    >>> plt.imshow(cy[:,:,1])
+    >>> plt.colorbar()
+    >>> plt.show()
+    >>> plt.imshow(cz[:,:,1])
+    >>> plt.colorbar()
+    >>> plt.show()
+    """
+    N = [0]*len(v)
```

Comment:
No, for divergence I felt you either should omit all sanity checks, since if the shapes did not match the calculation of the divergence would fail anyway, or just do the very simple one I gave, since that covers what you had anyway.
My note was just to say that for curl, in addition to the (possible) check that shapes are identical, you do need to check that `v` has length 3 (as you say too).